lib/srv/db/mysql/proxy.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/srv/db/mysql/protocol"
+
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/server"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// Proxy proxies connections from MySQL clients to database services
+// over reverse tunnel. It runs inside Teleport proxy service.
+//
+// Implements common.Proxy.
+type Proxy struct {
+	// TLSConfig is the proxy TLS configuration.
+	TLSConfig *tls.Config
+	// Middleware is the auth middleware.
+	Middleware *auth.Middleware
+	// Service is used to connect to a remote database service.
+	Service common.Service
+	// Log is used for logging.
+	Log logrus.FieldLogger
+}
+
+// HandleConnection accepts connection from a MySQL client, authenticates
+// it and proxies it to an appropriate database service.
+func (p *Proxy) HandleConnection(ctx context.Context, clientConn net.Conn) (err error) {
+	server := p.makeServer(clientConn)
+	// Perform first part of the handshake, up to the point where client sends
+	// us certificate and connection upgrades to TLS.
+	tlsConn, err := p.performHandshake(server)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	ctx, err = p.Middleware.WrapContextWithUser(ctx, tlsConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	serviceConn, err := p.Service.Connect(ctx, server.GetUser(), server.GetDatabase())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer serviceConn.Close()
+	// Before replying OK to the client which would make the client consider
+	// auth completed, wait for OK packet from db service indicating auth
+	// success.
+	err = p.waitForOK(server, serviceConn)
```

Comment:
Do you know if we impose timeouts/deadlines on the reverse tunnel connections?

Reply:
I don't know for sure, I think the tunnel stays open as long as heartbeats are coming. FWIW, I've added a deadline to this waitForOK call in case the other side misbehaves for some reason.